Add emission temperature threshold for sourcing particles per cell.

RyanWollaeger opened the pull request:
## Background

* Low / floor temperature regions may not need to source particles, depending on problem.
* Disabling sourcing in floored cells permits reallocating the user's target number of particles into cells above the temperature threshold, potentially improving statistics in those cells.

## Description of Changes

+ Add emit_temp_threshold input parameter.
+ Perform global reduction on total count of cells above threshold.
+ Use emitting cell count to adjust number per emitting cell up.
+ Set source numbers and energy weight to 0 for non-emitting cells.

## Checklist

- [ ] New features are documented
- [ ] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files

RyanWollaeger committed:
Add emission temperature threshold for sourcing particles per cell.

+ Add emit_temp_threshold input parameter.
+ Perform global reduction on total count of cells above threshold.
+ Use emitting cell count to adjust number per emitting cell up.
+ Set source numbers and energy weight to 0 for non-emitting cells.

RyanWollaeger commented:
Hm - thought the volta CI was launched - tried now launching again

RyanWollaeger committed:
Remove invalid if constexpr (that caused a CUDA runtime error).

Note: the source number mod applies to initial and time sources.

RyanWollaeger commented:
Marked this as draft - saw unexpected 0 active cells on Darwin GPU run - need to explore this more.

RyanWollaeger committed:
Add MPI_Allreduce option (argument) to global_sum_reduce.

RyanWollaeger commented on src/jaybenne/jaybenne_utils.hpp:
Does this incarnation of `par_reduce` sum over blocks?

RyanWollaeger replied:
Guessing using this for the `KOKKOS_LAMBDA` in `sourcing.cpp` is a problem, due to the lack of inlining decorators - going to try just using this form directly inline in `sourcing.cpp`.

RyanWollaeger replied:
Inlining wasn't the issue, evidently (still not sure what is). Trying a par_reduce_inner instead.

RyanWollaeger replied:
Found the issue `nblocks = vmesh.GetNBlocks()` is 1 while `md->GetMeshPointer()->nbtotal` is 100 for the problem, on one rank. This may be from an instantiation of the first argument to MeshBlockData rather than MeshData.

adamdempsey90 replied:
In artemis, we just do `md->NumBlocks()`

adamdempsey90 replied:
well, I guess you sometimes have this as `MeshBlockData`

RyanWollaeger replied:
Yes indeed - `MeshBlockData` gets used at initialization (via ProblemGenerator's), and is a templated argument for this routine.

RyanWollaeger replied:
Guess the workaround for now might be to restrict this to emission mode for now, where it makes most sense.

RyanWollaeger committed:
Thermal initialization only operates on single mesh blocks.

+ Remove superfluous instantiations of SourcePhotons.
+ Restrict temperature thresholding logic to emission mode.

RyanWollaeger commented:
With the latest change, this appears to be working now - testing on a few different ranks of the crooked pipe problem, the number of active emitting cells is the same. The drawback of the latest change is the initialization does not benefit from the particle redistribution, but this will be a small degradation for many problems.

alexrlongne commented:
LGTM, ready to merge?

RyanWollaeger commented:
> LGTM, ready to merge?

Thanks @alexrlongne! This should be ready (no changes planned)
